crates/ink/codegen/src/generator/metadata.rs

```diff
@@ -316,6 +354,48 @@ impl Metadata<'_> {
         }
     }
 
+    /// Generates ink! metadata for the given return type of a constructor.
+    fn generate_constructor_return_type(ret_ty: Option<&syn::Type>) -> TokenStream2 {
+        /// Returns `true` if the given type is `Self`.
+        fn type_is_self_val(ty: &syn::Type) -> bool {
+            matches!(ty, syn::Type::Path(syn::TypePath {
+            qself: None,
+            path
+        }) if path.is_ident("Self"))
+        }
+
+        match ret_ty {
+            None => {
+                quote! {
+                    ::ink::metadata::ReturnTypeSpec::new(::core::option::Option::None)
+                }
+            }
+            Some(ty) => {
+                if type_is_self_val(ty) {
+                    return quote! { ::ink::metadata::ReturnTypeSpec::new(::core::option::Option::None)}
+                }
+                let type_spec = Self::generate_constructor_type_spec(ty);
+                let type_token = Self::replace_self_with_unit(ty);
+                quote! {
+                    if !::ink::is_result_type!(#type_token) {
+                        ::ink::metadata::ReturnTypeSpec::new(::core::option::Option::None)
+                    } else {
+                       ::ink::metadata::ReturnTypeSpec::new(#type_spec)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This again does syntactical checks instead of using the Rust type system properly and will fail for various syntactic contexts. To do this properly you'd need to take a similar approach as in `generate_trait_messages` with Rust type system reflection such as
```rust
                let is_payable = quote! {{
                    <<::ink::reflect::TraitDefinitionRegistry<<#storage_ident as ::ink::reflect::ContractEnv>::Env>
                        as #trait_path>::__ink_TraitInfo
                        as ::ink::reflect::TraitMessageInfo<#local_id>>::PAYABLE
                }};
```
We do have the information for ink! messages and constructors as well. They are identified using their selectors which are always known at proc. macro compile time. Therefore you can access them in this scope just like in the linked example.

Reply:
The metadata codegen takes place after the WASM blob compilation, hence at this stage all types have already been evaluated by compiler and we can work with syntax.

In fact, I *generally* do not do syntactical checks. I only check if the token is `Self` here because metadata generation takes place outside the storage definition, hence `Self` can not be evaluated there. Same logic applies for `Result<Self, ...>`.

However, if the return type is the explicit type of a contract or `Result` with explicit contract type for `Ok` variant, then I do the type evaluation, hence why we have `TransformResult` factory.
I can be sure that the `Self` is valid in this context is because the WASM build is already done by now. 

I am not how message selectors are relevant here.

Reply:
I think we may be able to refactor this to make it use the Rust type system, I'm looking into this now.

Reply:
I've done an experiment with this in https://github.com/paritytech/ink/pull/1491, take a look and see what you think.